### ISSUE TITLE
Undefine them all!

### ISSUE THREAD

**Issue by nepalez** [dry-rb/dry-initializer#21]
Recently I recognized one "dirty" that the initializer has 2 different responsibilities. One is assigning variables, another is their validation.

IMO validation/coercion via `type:`, and assigning `default:` value seems ok.

But `optional: true` provides another layer of validation which seems totally excessive.
It either assigns `Dry::Initializer::UNDEFINED` or raises in case no value has been assigned.

And I thought, why don't we assign `Dry::Initializer::UNDEFINED` to any variable by default.

```ruby
class Foo
  extend Dry::Initializer::Mixin
  param :bar
end

# CURRENT behavior
Foo.new # => Boom!!!

# NEW behavior
Foo.new.instance_variable_get :@bar # => Dry::Initializer::UNDEFINED
```

Btw, with such a decision, you're sure you can add positional argument `param` at any time. Now you cannot add required `param` after the optional one, and this can provide tricky cases when subclassing.

```ruby
class Foo
  extend Dry::Initializer::Mixin
  param :bar, optional: true
end

class Bar < Foo
  param :baz # BOOM! (should be optional because it is positioned after :bar)
end
```

For those who still need that validation, we could keep `optional: false` (to make that boom when no value was provided).

Providing any declared `param` and `option` is `UNDEFINED` by default, then we could "hide" this value with a bit of magics (I know, I know...). Instead of using `attr_reader`, we could define readers in a way:

```ruby
def name
  @name == Dry::Initializer::UNDEFINED ? nil : @name
end
```

We still keep knowledge about whether a **variable** has been assigned, or not. At the same time we let the **reader** to behave like `nil` (which is another idiom for absent value in Ruby) to escape all that hacks processing `Dry::Initializer::UNDEFINED` when we don't care the difference.

To recap, I'd propose the following changes:

1) Assign `Dry::Initializer::UNDEFINED` when no other default value is provided (via `default: proc {...}`)
2) Make all arguments - both hash and positional - optional (and remove constraint for the order of positional ones)
3) Reload readers to hide `Dry::Initializer::UNDEFINED` values of the variables
4) Raise an exception in case of `optional: false` (or `required: true`) when no value was assigned explicitly

@solnic, @flash-gordon, @AMHOL, @marshall-lee wdyt?

**Comment by flash-gordon** [comment]
I'm not sure about this change, can you explain what's the problem with subclassing? Probably this can be fixed in a similar way to what we did in dry-auto_inject by using `Method#parameters` for building a proper `super` call (https://github.com/dry-rb/dry-auto_inject/tree/master/lib/dry/auto_inject/strategies).

P.S. Since dry-core is not a transitive dependency (via dry-types) you can use `Undefined` from there.

**Comment by solnic** [comment]
Actually, I think that using `UNDEFINED` as the default value is weird, because it forces us to use `proc { nil }` as the default value just because `UNDEFINED` is truthy, so you can't use it in conditionals.

**Comment by nepalez** [comment]
@solnic I feel you pain, yes. The problem is there're cases when we should carefully draw a difference between values that were set to `nil` explicitly, and those we do not care about. The most prominent one is data structures to be later merged to others (update persistence etc.). The business cases are different. In one case we are going to reset existing value, in the other - to keep it untouched.

`UNDEFINED` is just a "sort of" `nil` embellished with additional information about its origin.

But at the same time I agree to you that in most cases we don't care about such a difference, and `UNDEFINED` becomes an overkill. That is why I'd prefer to hide it behind a getter interface. Undefined attributes would quack as `nil`, but their underlying variables are still in a service to call us whether a variable was set explicitly, or not.

```ruby
class Foo
  extend Dry::Initializer::Mixin

  option :bar, optional: true
  option :baz, optional: true
end

foo = Foo.new bar: nil

foo.bar # => nil
foo.baz # => nil # MY PROPOSAL FOR v0.12.0 For now it returns UNDEFINED

foo.instance_variable_get :@bar # => nil (explicitly set to 'no value')
foo.instance_variable_get :@baz # => UNDEFINED
```

Another question is whether we should declare dependency from an argument potentially undefined (optional), or allow it explicitly. For now we have to explicitly mark either `param` or `option` with `optional: true`. I'd prefer to revert this design and forbid undefined values (assign `required: true`, or `optional: false`).

@flash-gordon tbh, the problem with reloading is not a "problem". It is a nature of positional argument, for which we cannot add an optional param after a required one. It is obvious in a parent class, but can be tricky in its subclass.

```ruby
class Foo
  extend Dry::Initializer::Mixin
  param :bar, optional: true
end

class Bar < Foo
  param :baz # WTF!!! This is because of the initializer(bar = UNDEFINED, baz)
end
```

We could skip that control at the load time (make any `param` UNDEFINED), and left it to runtime (like with hash arguments). But tbh I'm not sure this worths doing.

**Comment by nepalez** [comment]
Closed with reference to #22
